src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java

```diff
@@ -0,0 +1,116 @@
+import io.sphere.sdk.products.commands.updateactions.ChangeMasterVariant;
+import io.sphere.sdk.products.commands.updateactions.RemoveVariant;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.commercetools.sync.commons.utils.CollectionUtils.collectionToSet;
+import static com.commercetools.sync.commons.utils.CollectionUtils.filterCollection;
+import static com.commercetools.sync.commons.utils.CommonTypeUpdateActionUtils.buildUpdateAction;
+import static java.util.stream.Collectors.toList;
+
+public final class ProductVariantUpdateActionUtils {
+
+    /**
+     * <b>Note:</b> if you do both add/remove product variants - <b>first always remove the variants</b>,
+     * and then - add.
+     * The reason of such restriction: if you add first you could have duplication exception on the keys,
+     * which expected to be removed first.
+     *
+     * @param oldProduct old product with variants
+     * @param newProduct new product draft with variants <b>with resolved references prices references</b>
+     * @return list of update actions to remove missing variants.
+     * @see #buildAddVariantUpdateAction(Product, ProductDraft)
+     */
+    @Nonnull
+    public static List<UpdateAction<Product>> buildRemoveVariantUpdateAction(@Nonnull final Product oldProduct,
+                                                                             @Nonnull final ProductDraft newProduct) {
+        final List<ProductVariant> oldVariants = oldProduct.getMasterData().getStaged().getVariants();
+        final Set<String> newVariants = collectionToSet(newProduct.getVariants(),
+                ProductVariantDraft::getKey);
+
+        return filterCollection(oldVariants, oldVariant -> !newVariants.contains(oldVariant.getKey()))
+                .map(RemoveVariant::of)
+                .collect(toList());
+    }
+
+    /**
+     * <b>Note:</b> if you do both add/remove product variants - <b>first always remove the variants</b>,
+     * and then - add.
+     * The reason of such restriction: if you add first you could have duplication exception on the keys,
+     * which expected to be removed first.
+     *
+     * @param oldProduct old product with variants
+     * @param newProduct new product draft with variants <b>with resolved references prices references</b>
+     * @return list of update actions to add new variants.
+     * @see #buildRemoveVariantUpdateAction(Product, ProductDraft)
+     */
+    @Nonnull
+    public static List<UpdateAction<Product>> buildAddVariantUpdateAction(@Nonnull final Product oldProduct,
+                                                                          @Nonnull final ProductDraft newProduct) {
+        final List<ProductVariantDraft> newVariants = newProduct.getVariants();
+        final Set<String> oldVariantKeys =
+                collectionToSet(oldProduct.getMasterData().getStaged().getVariants(), ProductVariant::getKey);
```

Comment:
@andrii-kovalenko-ct `getVariants()` doesn't include the master variant. `getAllVariants()` does. So It means that the master variant will never be synced this way! 😮